lib/improver/wind_downscaling.py

```diff
@@ -0,0 +1,967 @@
+                try:
+                    z0_cube.remove_coord(unwanted_coord)
+                except Exception:
+                    pass
+            if z0_cube.units != Unit('m'):
+                msg = ("z0 ancil has unexpected unit: should be {} "
+                       "is {}")
+                raise ValueError(msg.format(Unit('m'), z0_cube.units))
+        mlist = list(itertools.permutations(alist, 2))
+        oklist = []
+        for entr in mlist:
+            oklist.append(entr[0].coords() == entr[1].coords())
+            # HybridHeightToPhenomOnPressure._cube_compatibility_check(entr[0],
+            # entr[1])
+        return np.array(oklist).all()  # replace by a return value of True
+
+    def find_coord_order(self, mcube):
+        """Extract coordinate ordering within a cube.
+
+        Figure out the order of the xyzt dimensions in a cube.
+        iris.cube.Cube.slices seems not always to return the order one
+        specifies, hence I need to find out the order and work with
+        transpose instead.
+
+        Parameters:
+        ----------
+        mcube: iris cube
+            cube to check the order of coordinate axis
+
+        Returns:
+        -------
+        xpos: integer
+            position of x axis.
+        ypos: integer
+            position of y axis.
+        zpos: integer
+            position of z axis.
+        tpos: integer
+            position of t axis.
+
+        """
+        xpos = ypos = zpos = tpos = np.nan
+        for iii, mcc in enumerate(mcube.coords()):
+            try:
+                if mcc == mcube.coord(self.x_name):
+                    xpos = iii
+                elif mcc == mcube.coord(self.y_name):
+                    ypos = iii
+                elif mcc == mcube.coord(self.z_name):
+                    zpos = iii
+                elif mcc == mcube.coord(self.t_name):
+                    tpos = iii
+            except Exception:
+                pass
+        return xpos, ypos, zpos, tpos
+
+    def find_heightgrid(self, wind):
+        """Setup the height grid.
+
+        Setup the height grid either from the 1D or 3D height grid
+        that was supplied to the plugin or from the z-axis information
+        from the wind grid.
+
+        Parameters:
+        ----------
+        wind: 3D or 4D iris cube
+            representing the wind data.
+
+        Returns:
+        -------
+        hld: 1D or 3D np.array
+            representing the height grid.
+
+        """
+        if self.height_levels is None:
+            hld = wind.coord(self.z_name).points
+        else:
+            hld = iris.util.squeeze(self.height_levels)
+            if np.isnan(hld.data).any() or (hld.data == RMDI).any():
+                raise ValueError("height grid contains invalid points")
+            if hld.ndim == 3:
+                try:
+                    xap, yap, zap, _ = self.find_coord_order(hld)
+                    hld.transpose([yap, xap, zap])
+                except Exception:
```

Comment:
Catching KeyError caused by using transpose on dimensions that don't exist e.g. hld.transpose([4, 5, 6]), if hld is a 3d cube.